### PR TITLE
prometheusremotewrite: release lock when wal empty

### LINKED ISSUE #15277: [exporter/prometheusremotewrite] Enabling WAL prevents metrics from being forwarded
## Description
When using the prometheusremotewrite exporter with the WAL enabled, no metrics are sent from the collector to the remote write destination.

## Steps to Reproduce
Using the config in the config section below can reproduce this error by sending metrics to this collector. Disabling the WAL section causes all metrics to be sent properly.

## Expected Result
Prometheus metrics should appear in the remote write destination.

## Actual Result
No metrics were sent to the remote write destination.

### Collector version

0.62.1

## Environment
AWS bottlerocket running otel/opentelemetry-collector-contrib:0.36.3 docker image

### OpenTelemetry Collector configuration

```yaml
exporters:
  logging:
    loglevel: info
  prometheusremotewrite:
    endpoint: http://thanos-receive-distributor:19291/api/v1/receive
    remote_write_queue:
      enabled: true
      num_consumers: 1
      queue_size: 5000
    resource_to_telemetry_conversion:
      enabled: true
    retry_on_failure:
      enabled: false
      initial_interval: 5s
      max_elapsed_time: 10s
      max_interval: 10s
    target_info:
      enabled: false
    timeout: 15s
    tls:
      insecure: true
    wal:
      buffer_size: 100
      directory: /data/prometheus/wal
      truncate_frequency: 45s
extensions:
  health_check: {}
  memory_ballast: {}
  pprof:
    endpoint: :1888
  zpages:
    endpoint: :55679
processors:
  batch: {}
  batch/metrics:
    send_batch_max_size: 500
    send_batch_size: 500
    timeout: 180s
  memory_limiter:
    check_interval: 5s
    limit_mib: 4915
    spike_limit_mib: 1536
receivers:
  jaeger:
    protocols:
      grpc:
        endpoint: 0.0.0.0:14250
      thrift_compact:
        endpoint: 0.0.0.0:6831
      thrift_http:
        endpoint: 0.0.0.0:14268
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
  prometheus:
    config:
      scrape_configs:
      - job_name: opentelemetry-collector
        scrape_interval: 10s
        static_configs:
        - targets:
          - ${MY_POD_IP}:8888
  zipkin:
    endpoint: 0.0.0.0:9411
service:
  extensions:
  - health_check
  - pprof
  - zpages
  pipelines:
    logs:
      exporters:
      - logging
      processors:
      - memory_limiter
      - batch
      receivers:
      - otlp
    metrics:
      exporters:
      - prometheusremotewrite
      processors:
      - batch/metrics
      receivers:
      - otlp
    traces:
      exporters:
      - logging
      processors:
      - memory_limiter
      - batch
      receivers:
      - otlp
      - jaeger
      - zipkin
  telemetry:
    metrics:
      address: 0.0.0.0:8888
```

### Log output

_No response_

### Additional context

From debugging, this looks to be a deadlock between `persistToWAL()` and `readPrompbFromWAL()`, but I'm not 100% certain

### LINKED ISSUE #19363: exporter/prometheusremotewrite: wal leads to oom under high load
**Describe the bug**
When running the `prometheusremotewrite` exporter in wal enabled mode under (very) high load (250k active series), it quickly builds up memory until the kernel oom kills otelcol

**Steps to reproduce**

<details>
<summary><code>docker-compose.yml</code></summary>

```yml
version: '2'
services:
  # generates 250k series to be scaped by otelcol
  avalanche:
    image: quay.io/prometheuscommunity/avalanche:main
    command:
      - --metric-count=1000
      - --series-count=250
      - --label-count=5
      - --series-interval=3600
      - --metric-interval=3600

  otel:
    image: otel/opentelemetry-collector
    volumes: [otel-cfg:/etc/otelcol]
    user: 0:0
    tmpfs:
      - /wal
    depends_on:
      otel-cfg:
        condition: service_completed_successfully
    mem_limit: 8G
    restart: always

  otel-cfg:
    image: alpine
    volumes: [otel-cfg:/etc/otelcol]
    command:
      - sh
      - -c
      - |
        cat - > /etc/otelcol/config.yaml << EOF
        receivers:
          prometheus:
            config:
              scrape_configs:
              - job_name: stress
                scrape_interval: 15s
                static_configs:
                  - targets:
                    - avalanche:9001
        processors:
          batch:
        exporters:
          prometheusremotewrite:
            endpoint: http://receiver:9090/api/v1/write
            wal:
              directory: /wal
        service:
          pipelines:
            metrics:
              receivers: [prometheus]
              processors: [batch]
              exporters: [prometheusremotewrite]
        EOF
  # dummy http server to "receive" remote_write samples by always replying with http 200
  receiver:
    image: caddy
    command: sh -c 'echo ":9090" > /tmp/Caddyfile && exec caddy run --config /tmp/Caddyfile'

  # prometheus observing resource usage of otelcol
  prometheus:
    image: prom/prometheus
    ports:
      - 9090:9090
    entrypoint: /bin/sh
    command:
      - -c
      - |
        cat - > prometheus.yml << EOF && /bin/prometheus
        global:
          scrape_interval: 5s
        scrape_configs:
          - job_name: otel
            static_configs:
              - targets:
                - otel:8888
        EOF
volumes:
  otel-cfg: {}
```

</details>

**What did you expect to see?**
Otelcol having a (high) but (periodically) stable memory usage

**What did you see instead?**

Otelcol repeatedly builds up memory until it is oom killed by the operating system, only to repeat this exact behavior

Screenshot from 2023-03-06 23-33-13

**What version did you use?**
Version: Docker `otel/opentelemetry-collector-contrib:0.72.0`

**What config did you use?**
See above `docker-compose.yml`

**Environment**

<details>
<summary><code>docker info</code></summary>

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc., 2.13.0)

Server:
 Containers: 31
  Running: 0
  Paused: 0
  Stopped: 31
 Images: 65
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 770bd0108c32f3fb5c73ae1264f7e503fe7b2661.m
 runc version: 
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.12.10-arch1-1
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 15.39GiB
 Name: <omit>
 ID: <omit>
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: <omit>
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

</details>

**Additional context**
This only occurs when enabling wal mode. Other prometheus agents (Grafana Agent, Prometheus Agent Mode) do not show this behavior on the exact same input data

### PR DESCRIPTION
**Pull request by tombrk**
**Description:** <Describe what has changed.>

Previously, the wal tailing routine held an exclusive lock over the entire wal, while waiting for new sections to be written.

This led to a deadlock situtation, as a locked wal naturally can not be written to.

To remedy this, the lock is now only held during actual read attempts, not in between.

Furthermore, inotify watching of the wal dir has been replaced with channel-based signaling between the producers and the tailing routine, so that the latter blocks until any new writes have happened.

**Link to tracking Issue:**

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19363
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/15277

**Testing:**

An "end to end" test has been added where the WAL is first initialized without any initial data, so it goes into waiting mode. Writes are performed afterwards and the read-back result is verified.

This previously led to a deadlock and now no longer does.

It has also been confirmed that goroutines no longer build up and actual http requests are now being made.

### CODE REVIEW

File: exporter/prometheusremotewriteexporter/wal.go [hunk 1]
```diff
@@ -324,113 +332,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 			return err
 		}
 		wIndex := prwe.wWALIndex.Add(1)
+		prwe.log.Debug("write", zap.Uint64("index", wIndex))
```
Comment (frzifus) [inline exporter/prometheusremotewriteexporter/wal.go:335]: Do we still need it or can we set it to trace level? (in that case, it should be a bit more explicit)

File: exporter/prometheusremotewriteexporter/wal.go [hunk 2]
```diff
@@ -324,113 +332,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
...
+	}
 
-			// Now increment the WAL's read index.
-			prwe.rWALIndex.Add(1)
+	try := func() (*prompb.WriteRequest, error) {
+		prwe.mu.Lock()
+		defer prwe.mu.Unlock()
 
-			return req, nil
+		prwe.log.Debug("read", zap.Uint64("index", index))
```
Comment (frzifus) [inline exporter/prometheusremotewriteexporter/wal.go:364]: same here

File: exporter/prometheusremotewriteexporter/wal.go [hunk 3]
```diff
@@ -324,113 +332,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
...
-		go func() {
-			defer func() {
-				watchCh <- wErr
-				close(watchCh)
-				// Close the file watcher.
-				walWatcher.Close()
-			}()
+		req, err := try()
+		if errors.Is(err, wal.ErrNotFound) {
+			prwe.log.Debug("wal empty - waiting for write")
```
Comment (frzifus) [inline exporter/prometheusremotewriteexporter/wal.go:390]: and here

File: exporter/prometheusremotewriteexporter/wal.go (hunk 1 above)
Comment (sh0rez) [reply exporter/prometheusremotewriteexporter/wal.go:335]: turns out – there is no `trace` level in zap ... so we need to either leave it as debug or remove entirely

Comment (frzifus) [reply exporter/prometheusremotewriteexporter/wal.go:335]: how is the full log message displayed?

**Comment by frzifus** [comment]
Whats the state here? :) ping @sh0rez

**Review by frzifus — APPROVED** [review]

File: exporter/prometheusremotewriteexporter/wal.go (hunk 1 above)
Comment (sh0rez) [reply exporter/prometheusremotewriteexporter/wal.go:335]:
```
=== RUN   TestWAL_E2E
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 1}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:390	wal empty - waiting for write
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:335	write	{"index": 1}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:335	write	{"index": 2}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:335	write	{"index": 3}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 1}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 2}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 3}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 4}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:390	wal empty - waiting for write
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:364	read	{"index": 4}
2023-05-25T16:28:08.513+0200	DEBUG	prometheusremotewriteexporter/wal.go:390	wal empty - waiting for write
```

**Comment by codeboten** [comment]
Looks like the checks hung, re-opening to trigger the checks

**Review by codeboten — COMMENTED** [review]
Please add a changelog entry for the bug fix

**Review by codeboten — COMMENTED** [review]
Looks like a test is failing, please have a look

```
--- FAIL: TestWAL_persist (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa696b6]

goroutine 292 [running]:
testing.tRunner.func1.2({0x11fa360, 0x1abbe70})
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1526 +0x372
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1529 +0x650
panic({0x11fa360, 0x1abbe70})
	/opt/hostedtoolcache/go/1.20.4/x64/src/runtime/panic.go:890 +0x263
go.uber.org/zap.(*Logger).check(0x0, 0xff, {0x1306013, 0x5})
	/home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:298 +0x76
go.uber.org/zap.(*Logger).Debug(0x149d400?, {0x1306013, 0x5}, {0xc0003cb2c0, 0x1, 0x1})
	/home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:211 +0x55
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prweWAL).persistToWAL(0xc0000a60e0, {0xc000175e20, 0x2, 0x0?})
	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter/wal.go:324 +0x308
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.TestWAL_persist(0xc0001f4d00)
	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter/wal_test.go:138 +0xb11
testing.tRunner(0xc0001f4d00, 0x135cf28)
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.20.4/x64/src/testing/testing.go:1629 +0x806
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter	0.074s
FAIL
```

**Comment by sh0rez** [comment]
@codeboten any chance you could take another look?

**Comment by frzifus** [comment]
ping @sh0rez

File: exporter/prometheusremotewriteexporter/wal.go [hunk 4]
```diff
@@ -66,13 +68,17 @@ func newWAL(walConfig *WALConfig, exportSink func(context.Context, []*prompb.Wri
...
-	return &prweWAL{
+	wal := prweWAL{
 		exportSink: exportSink,
 		walConfig:  walConfig,
 		stopChan:   make(chan struct{}),
+		rNotify:    make(chan struct{}),
 		rWALIndex:  &atomic.Uint64{},
 		wWALIndex:  &atomic.Uint64{},
-	}, nil
+		log:        zap.NewNop(),
```
[the remark this answers is not in the document]
Comment (frzifus) [reply exporter/prometheusremotewriteexporter/wal.go:78]: There is a logger available in the `TelemetrySettings`. But adding a `zap.NewNop()` logger here, should be equal to the state before without any logger, right?

Not that I have a strong opinion about it - I just would like to resolve it at some point :smile:

**Review by frzifus — COMMENTED** [review]
@sh0rez I created a small [PR](https://github.com/sh0rez/opentelemetry-collector-contrib/pull/1) to your branch to address the requested changes.

File: exporter/prometheusremotewriteexporter/wal.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply exporter/prometheusremotewriteexporter/wal.go:78]: I think it is more the point of, if we are adding a logger, we need some way of providing a configuration for it.

File: exporter/prometheusremotewriteexporter/wal.go [hunk 5]
```diff
@@ -313,113 +322,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
...
-			if err = proto.Unmarshal(protoBlob, req); err != nil {
-				return nil, err
-			}
+	if prwe == nil {
+		return nil, fmt.Errorf("attempt to read from closed WAL")
+	}
 
-			// Now increment the WAL's read index.
-			prwe.rWALIndex.Add(1)
+	try := func() (*prompb.WriteRequest, error) {
```
Comment (MovieStoreGuy) [inline exporter/prometheusremotewriteexporter/wal.go:350]: Could I ask that you move this to be a method as part of the type?

My concern being is that trying to unit test this means a lot of potential set up and keeping things simple and isolated helps checking where things break (or making sure they break correctly).

File: exporter/prometheusremotewriteexporter/wal.go [hunk 6]
```diff
@@ -313,113 +322,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
...
-			return nil, werr
-		}
-		if werr = walWatcher.Add(prwe.walPath); werr != nil {
-			return nil, werr
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-prwe.stopChan:
+			return nil, fmt.Errorf("attempt to read from WAL after stopped")
```
Comment (MovieStoreGuy) [inline exporter/prometheusremotewriteexporter/wal.go:374]: I would suggest using `errors.New` instead of `fmt.Errorf` since it should be "cheaper" to create the value since it doesn't need to parse the string for any formatting symbols.

Alternatively, since this is static you can define the error at the type of the file as part of the "static" vars.

File: exporter/prometheusremotewriteexporter/wal_test.go [hunk 7]
```diff
@@ -101,6 +103,7 @@ func TestWAL_persist(t *testing.T) {
 
 	pwal, err := newWAL(config, doNothingExportSink)
 	require.Nil(t, err)
+	pwal.log = zap.Must(zap.NewDevelopment())
```
Comment (MovieStoreGuy) [inline exporter/prometheusremotewriteexporter/wal_test.go:106]: To make your life easier, I suggest using `zaptest.NewLogger(t)` within tests so that you get logs only if the tests fails.

File: exporter/prometheusremotewriteexporter/wal_test.go [hunk 8]
```diff
@@ -156,3 +159,64 @@ func TestWAL_persist(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(t.TempDir())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log, err := zap.NewDevelopment()
+	if err != nil {
+		panic(err)
+	}
```
Comment (MovieStoreGuy) [inline exporter/prometheusremotewriteexporter/wal_test.go:193]: You shouldn't need another logger here, I suggest you create one at the start of the test and use that throughout.

File: exporter/prometheusremotewriteexporter/wal_test.go [hunk 9]
```diff
@@ -156,3 +159,64 @@ func TestWAL_persist(t *testing.T) {
...
+	ctx = contextWithLogger(ctx, log)
+
+	if err := wal.run(ctx); err != nil {
+		panic(err)
```
Comment (MovieStoreGuy) [inline exporter/prometheusremotewriteexporter/wal_test.go:197]: Please use `require.NoError` instead of panicing.

**Review by MovieStoreGuy — COMMENTED** [review]
Just a few things from me.

File: exporter/prometheusremotewriteexporter/wal.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sh0rez) [reply exporter/prometheusremotewriteexporter/wal.go]: the logger is taken from the context once `run()` is called, but we need a place to store it for long running operations.

not setting this var leads to a nil pointer that panics, so its `NewNop()` instead to avoid that. afaict this is never called before run() anways

**Comment by sh0rez** [comment]
@frzifus @Aneurysm9 @codeboten somehow dropped the ball on this, but it's still relevant.

rebased and addressed review comments, please take another look!

**Review by jpkrohling — APPROVED** [review]
LGTM

**Review by frzifus — APPROVED** [review]
Thx, for picking this up again. :)

File: exporter/prometheusremotewriteexporter/wal_test.go [hunk 10]
```diff
@@ -149,3 +152,57 @@ func TestWAL_persist(t *testing.T) {
...
+	if err := wal.run(ctx); err != nil {
+		require.NoError(t, err)
+	}
+
+	if err := wal.persistToWAL(in); err != nil {
+		require.NoError(t, err)
+	}
+
+	// wait until the tail routine is no longer busy
+	wal.rNotify <- struct{}{}
```
Comment (jpkrohling) [inline exporter/prometheusremotewriteexporter/wal_test.go:191]: Just to confirm: if I remove this, the test will be blocked on the `<-done` until it times out. That's the purpose of this test, and what you are fixing here, right?

Comment (Aneurysm9) [reply exporter/prometheusremotewriteexporter/wal_test.go:191]: I'm also a bit confused as to why this is needed here.  `run()` will fork and (indirectly) call `read()`, blocking until it returns, which will happen when a record is read from the WAL or the context is cancelled.  `read()` will try to read from the WAL, find nothing, and then block reading from `rNotify`.  `persistToWAL()` will then acquire the mutex, signal on `rNotify`, and write to the WAL.  Once `persistToWAL()` signals on `rNotify` then `read()` starts blocking on acquiring the mutex in `tryRead()`, eventually acquiring it once `persistToWAL()` executes its deferred unlock.  `read()` will then get a successful result from `tryRead()`, popping all the way back out to `continuallyPopWALThenExport()` which may or may not actually export (probably not with the default truncate frequency and buffer size), and then goes right back into `read()`.  At this point the WAL is empty and `tryRead()` returns the sentinel error `read()` uses to decide to wait on `rNotify`.  Since `persistToWAL()` has returned nothing is going to send anything on `rNotify` until this line here.  Sending a value here causes `read()` to try again, but nothing happens.  Then the context is cancelled and a bunch of `select` statements hit their `<-ctx.Done()` case, causing `read()` to return `nil, context.Canceled` to `continuallyPopWALThenExport()`, which returns `context.Canceled` to `run()`, executing its deferred export along the way, causing `sink` to close `done` and allowing this function to reach its requirement that the exported data equals the input data.

tl;dr: I don't see what writing to this channel does that isn't also accomplished by cancelling the context.

File: exporter/prometheusremotewriteexporter/wal.go [hunk 11]
```diff
@@ -70,8 +72,12 @@ func newWAL(walConfig *WALConfig, exportSink func(context.Context, []*prompb.Wri
 		exportSink: exportSink,
 		walConfig:  walConfig,
 		stopChan:   make(chan struct{}),
+		rNotify:    make(chan struct{}),
 		rWALIndex:  &atomic.Uint64{},
 		wWALIndex:  &atomic.Uint64{},
+
+		// populated from context in run()
+		log: zap.NewNop(),
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply exporter/prometheusremotewriteexporter/wal.go:80]: I had the exact same feeling, but thought it would be ok to have it like that given it kind of follows the pattern of the existing code. I'd be ok to have this as part of a follow up PR.

[the remark this answers is not in the document]
Comment (jpkrohling) [reply exporter/prometheusremotewriteexporter/wal.go:80]: I had the exact same feeling, but thought it would be ok to have it like that given it kind of follows the pattern of the existing code. I'd be ok to have this as part of a follow up PR.

[the remark this answers is not in the document]
Comment (Aneurysm9) [reply exporter/prometheusremotewriteexporter/wal.go:80]: I'm torn on this.  I feel strongly that we shouldn't be knowingly introducing new problems when we also know what the fix is, particularly when the fix isn't a big lift.  I also realize this is a purely internal change and doesn't expose any of the wrongness outside of the package.

The other conversation about the blocking on the notify channel in the test seems more important to me to resolve.  If we come to a resolution on that but this isn't fixed I think I'd be at a position where I wouldn't approve this PR but wouldn't object to it being merged with an understanding that this needs to be fixed as a fast follower.

File: exporter/prometheusremotewriteexporter/wal_test.go (hunk 10 above)
Comment (jpkrohling) [reply exporter/prometheusremotewriteexporter/wal_test.go:191]: @sh0rez , are you still interested in pursuing this?
